File: 프로그래머스/lv3/43162. 네트워크/네트워크.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> adj;
bool visited[200];

void DFS(int i, int n) {
    visited[i] = true;
    for (int j = 0; j < n; j++) {
        if (visited[j])
            continue;
        
        if (adj[i][j] == 1 || adj[j][i] == 1)
            DFS(j, n);
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    adj = computers;
    
    for (int i = 0; i < n; i++) {
        if (visited[i])
            continue;
        DFS(i, n);
        answer++;
    }
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.cs
using System;

public class Solution 
{
    private bool[] _visited;
    private int _answer = 0;
    
    public int solution(int n, int[,] computers) 
    {
        // 초기화
        _visited = new bool[n];
        for (int i = 0; i < n; i++)
            _visited[i] = false;
        ////
        
        for (int start = 0; start < n; start++)
        {
            if (DFS(start, n, computers) == true)
                _answer++;
        }
        
        return _answer;
    }
    
    // true: DFS 할 수 있음
    // false: DFS 할 수 없음
    public bool DFS(int now, int n, int[,] computers)
    {
        if (_visited[now] == true)
            return false;
        
        _visited[now] = true;
        
        for (int next = 0; next < n; next++)
        {
            if (now == next)
                continue;
            if (computers[now, next] == 0)
                continue;
            if (_visited[next] == true)
                continue;
            
            DFS(next, n, computers);
        }
        return true;
    }
}